### ISSUE TITLE
SWAT 4 - Player screams wiggle with HRTF enabled when the player is moving. Dead zone option for near listener sounds?

### ISSUE THREAD

**Issue by rkkoszewski** [kcat/openal-soft#522]
Hi, 
In SWAT 4 whenever the player is screaming and moving at the same time, when HRTF is enabled in openal-soft the scream is wiggling around the player head in a very noticeable way. When the player is still, the screams sound relatively fine. To me it sounds as if the scream audio is at the listener position, and when the player is moving eider the listener or the sound position are updated at different times or are following some grid, resulting in this wiggly sound. With HRTF disabled the player scream sounds perfectly fine no matter if the player is moving or not.

You can hear it here: https://youtu.be/SOYYkuZfn8s?t=112

Is there any way to add an optional dead-zone for when the listener is close to the audio source? Or have the HRTF effect stronger or less stronger, depending on the distance of the listener? Or could the issue be somewhere else?

**Comment by rkkoszewski** [comment]
Follow up on this one too, I added a deadzone in the "Source to Listener" step (Hardcoded by now) in alu.cpp. This seems to fix the wobbly voice issue in SWAT 4 when using HRTF and there don't seem to be any audible artifacts coming from the change. 

```
...
const bool directional{Direction.normalize() > 0.0f};
alu::Vector ToSource{ Position[0], Position[1], Position[2], 0.0f };

// Source to Listener Deadzone
const float srclsnr_deadzone = 0.001f ;
if (Position[0] < srclsnr_deadzone && Position[0] > -srclsnr_deadzone && 
    Position[1] < srclsnr_deadzone && Position[1] > -srclsnr_deadzone && 
    Position[2] < srclsnr_deadzone && Position[2] > -srclsnr_deadzone) {
    ToSource = { 0.0f, 0.0f, 0.0f, 0.0f };
}

const float Distance{ToSource.normalize()};

/* Initial source gain */
GainTriplet DryGain{props->Gain, 1.0f, 1.0f};
GainTriplet WetGain[MAX_SENDS];
...
```

It's not really a fix, as this should be fixed in the game itself. But given that it's unlikely that the older games like SWAT 4 will ever get an update, especially taking into account that HRTF is not supported officially by the game, it would be great if there would be a "misc" option in softal.ini that would allow to specify a listener to source deadzone when necessary. Would such an addition be acceptable for OpenAL-Soft?

**Comment by kcat** [comment]
Having a kind of deadzone for sources that are 'close enough' to the listener is a good idea, but It's something that needs to be handled carefully since there's no telling what unit scale the app may be using. It could have `AL_REFERENCE_DISTANCE` set to `0.001f` or less, perhaps using 1 unit = 1km while handling stuff at the meter scale or less. A good deadzone for one app may be too large for another, and too small of something else. Floating point is flexible like that, retaining 24 bits of precision while scaling up and down using a 7-bit exponent. Given floating point inaccuracies, the amount of error that seeps into the result will be scaled according to the size of the values used.

**Comment by kcat** [comment]
I've added a bit more of a flexible deadzone with commit 3702ee669b893ceb6ef7571c9f2134129af95dbe. Does that help stablize the position of sounds coming from the player while moving?

**Comment by rkkoszewski** [comment]
@kcat Yes! Huge difference! The wiggling seems to be completely gone now or is imperceptible by me. 
Thank you for looking into the issue and for the fix! 
I'll close the issue as it seems to be mitigated by the mentioned PR, feel free to reopen it if necessary.

**Comment by kcat** [comment]
Out of curiosity, is this still fixed with the latest commit (58a9549a5886fe91aba698abb4e6e4c0c94f129e)? I had to revert the variable deadzone because it was interacting poorly with another app, but I also modified handling the listener position so it should be a bit more stable overall.
